pylabrobot/storage/inheco/incubator_shaker_backend.py

```diff
@@ -0,0 +1,1875 @@
+      self.unit_composition[unit_index] = inc_type
+
+      await self.initialize(stack_index=unit_index)
+
+      self.loading_tray_status[unit_index] = await self.request_drawer_status(
+        stack_index=unit_index
+      )
+
+      self.temperature_control_status[unit_index] = await self.is_temperature_control_enabled(
+        stack_index=unit_index
+      )
+
+      if "shaker" in inc_type:
+        self.shaking_status[unit_index] = await self.is_shaking_enabled(stack_index=unit_index)
+      else:
+        self.shaking_status[unit_index] = False
+
+    self.setup_finished = True
+
+    msg = (
+      f"Connected to Inheco Incubator Shaker Stack on {self.port}\n"
+      f"DIP switch ID of bottom unit: {self.dip_switch_id}\n"
+      f"Number of connected units: {self.number_of_connected_units}\n"
+      f"Unit composition: {self.unit_composition}"
+    )
+    if verbose:
+      print(msg)
+    self._log(logging.INFO, msg)
+
+  async def stop(self):
+    """Close serial connection."""
+
+    for unit_index in range(self.number_of_connected_units):
+      if self.temperature_control_status[unit_index]:
+        await self.stop_temperature_control(stack_index=unit_index)
+
+      if self.shaking_status[unit_index]:
+        await self.stop_shaking(stack_index=unit_index)
+
+    if self.io:
+      await self.io.stop()
+      self.io = None
+
+  # === Low-level I/O ===
+
+  async def write(self, data: bytes) -> None:
+    """Write binary data to the serial device."""
+    self._log(logging.DEBUG, f"→ {data.hex(' ')}")
+    await self.io.write(data)
+
+  async def _read_full_response(self, timeout: float) -> bytes:
+    """Read a complete INHECO response frame asynchronously."""
+    if not self.io:
+      raise RuntimeError("Serial port not open.")
+
+    loop = asyncio.get_event_loop()
+    start = loop.time()
+    buf = bytearray()
+    expected_hdr = (0xB0 + self.dip_switch_id) & 0xFF
+
+    def has_complete_tail(b: bytearray) -> bool:
+      # Valid frame ends with: [hdr][0x20-0x2F][0x60]
+      return len(b) >= 3 and b[-1] == 0x60 and b[-3] == expected_hdr and 0x20 <= b[-2] <= 0x2F
+
+    while True:
+      chunk = await self.io.read(16)
+      if chunk:
+        buf.extend(chunk)
+        self._log(logging.DEBUG, chunk.hex(" "), direction="←")
+        if has_complete_tail(buf):
+          return bytes(buf)
+
+      if loop.time() - start > timeout:
+        raise TimeoutError(f"Timed out waiting for complete response (so far: {buf.hex(' ')})")
+
+      await asyncio.sleep(0.005)
+
+  # === Encoding / Decoding ===
+
+  def _crc8_legacy(self, data: bytearray) -> int:
+    """Compute legacy CRC-8 used by INHECO devices."""
+    crc = 0xA1
+    for byte in data:
+      d = byte
+      for _ in range(8):
+        if (d ^ crc) & 1:
+          crc ^= 0x18
+          crc >>= 1
+          crc |= 0x80
+        else:
+          crc >>= 1
+        d >>= 1
+    return crc & 0xFF
```

Comment:
I don't think there's a good pattern in PLR right now but ideally we share this with other inheco implementations